node/staking.go

```diff
@@ -32,25 +31,68 @@ func (node *Node) UpdateStakingList(stakeInfoReturnValue *structs.StakeInfoRetur
 		return
 	}
 	node.CurrentStakes = make(map[common.Address]*structs.StakeInfo)
+	node.CurrentStakesByNode = make(map[[20]byte]*structs.StakeInfo)
+	utils.GetLogInstance().Info("UpdateStakingList",
+		"numStakeInfos", len(stakeInfoReturnValue.LockedAddresses))
 	for i, addr := range stakeInfoReturnValue.LockedAddresses {
 		blockNum := stakeInfoReturnValue.BlockNums[i]
+		blsAddr := stakeInfoReturnValue.BlsAddresses[i]
 		lockPeriodCount := stakeInfoReturnValue.LockPeriodCounts[i]
-
 		startEpoch := core.GetEpochFromBlockNumber(blockNum.Uint64())
 		curEpoch := core.GetEpochFromBlockNumber(node.blockchain.CurrentBlock().NumberU64())
 
-		if startEpoch == curEpoch {
+		if startEpoch >= curEpoch {
```

Comment:
any comment on this change?

Reply:
This is for logical consistency/futureproofing.  `startEpoch` is the epoch in which the staking transaction was received.  The stake is valid from `(startEpoch + 1)` through `(startEpoch + lockPeriod)`; if `startEpoch` is equal to or greater than `curEpoch`, the stake is not yet active.

The current logic only captures `startEpoch == curEpoch` case, which is fine for now because a stake always begins to take effect in the immediately following epoch (`curEpoch + 1`), so `startEpoch` is always `curEpoch` when a stake is received.  However, if we relax the rule so that a stake can start taking effect farther down the road, the equality check will fail to catch such future stakes.  `startEpoch >= curEpoch` captures both cases.  Cc: @rlan35